pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,211 @@
+
+	// updateOnDNSMsg is a function to update the DNS message in the cilium agent on receiving the FQDN mapping
+	updateOnDNSMsg messagehandler.DNSRequestHandler
+
+	// snapshotMutex is a mutex to protect the current state of the DNS rules
+	snapshotMutex lock.Mutex
+
+	// identityToIpMutex is a mutex to protect the current state of the identity to Ip mapping
+	identityToIpMutex lock.Mutex
+
+	// currentIdentityToIp is a map of the identity to list of Ips
+	currentIdentityToIp map[identity.NumericIdentity][]net.IP
+
+	// log is the logger for the FQDNDataServer
+	log *slog.Logger
+}
+
+var (
+	kaep = keepalive.EnforcementPolicy{
+		PermitWithoutStream: true, // Allow pings even when there are no active streams
+	}
+	kasp = keepalive.ServerParameters{
+		Time:    5 * time.Second, // Ping the client if it is idle for 5 seconds to ensure the connection is still active
+		Timeout: 1 * time.Second, // Wait 1 second for the ping ack before assuming the connection is dead
+	}
+)
+
+type PolicyUpdater interface {
+	UpdatePolicyRulesLocked(map[identity.NumericIdentity]policy.SelectorPolicy, bool) error
+}
+
+// StreamPolicyState is a bidirectional streaming RPC to subscribe to DNS policies
+// SDP calls this method to subscribe to DNS policies
+// For each stream, we start a goroutine to receive the DNS policies ACKs
+// The flow of the method is as follows:
+// 1. Add the stream to the map( called by the client i.e SDP)
+// 2. Start a goroutine to receive the DNS policies ACKs for that particular client.
+// 3. Send the current state of the DNS rules to the client (We store the current state fo DNS rules during the endpoint regeneration see UpdatePolicyRulesLocked)
+// 4. Wait for the context to be done
+func (s *FQDNDataServer) StreamPolicyState(stream pb.FQDNData_StreamPolicyStateServer) error {
+	stream.Send(&pb.PolicyState{})
+	return nil
+}
+
+// NewServer creates a new FQDNDataServer which is used to handle the Standalone DNS Proxy grpc service
+func NewServer(endpointManager endpointmanager.EndpointManager, updateOnDNSMsg messagehandler.DNSRequestHandler, port int, logger *slog.Logger) *FQDNDataServer {
+
+	s := &FQDNDataServer{
+		port:                port,
+		endpointManager:     endpointManager,
+		updateOnDNSMsg:      updateOnDNSMsg,
+		currentIdentityToIp: make(map[identity.NumericIdentity][]net.IP),
+		log:                 logger.With(logfields.LogSubsys, "fqdn/server"),
+	}
+
+	return s
+}
+
+// OnIPIdentityCacheChange is a method to receive the IP identity cache change events
+func (s *FQDNDataServer) OnIPIdentityCacheChange(modType ipcache.CacheModification, cidr types.PrefixCluster, oldHostIP, newHostIP net.IP, oldID *ipcache.Identity, newID ipcache.Identity, encryptKey uint8, k8sMeta *ipcache.K8sMetadata, endpointFlags uint8) {
+	s.identityToIpMutex.Lock()
+	defer s.identityToIpMutex.Unlock()
+
+	if cidr.ClusterID() != 0 {
+		s.log.Debug("Ignoring IP identity cache change for non-local CIDR", logfields.CIDR, cidr.String())
```

Comment:
This says `Ignoring` but doesn't ignore

Reply:
https://github.com/cilium/cilium/pull/38669#discussion_r2027925994 With reference to this comment I want to make sure that we log the scenarios of `clusterID !=0` and `non single IP`. The actual addition/deletion from the ip cache is based on the `clusterId == 0` and `single Ip` so it is ignoring the clusterId != 0 and non single Ip.

For now I am removing the logs as it is counter intuitive to understand the code flow.